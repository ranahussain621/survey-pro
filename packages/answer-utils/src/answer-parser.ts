import { CURRENCY_SYMBOLS } from './consts'
import { Answer, FullNameValue, ServerSidePaymentValue } from '@heyform-inc/shared-types-enums'
import { helper } from '@heyform-inc/utils'
import Big from 'big.js'

function fileUpload(answer: Answer, livePreview = false) {
  if (livePreview) {
    return {
      filename: (answer as any).name
    }
  }

  if (helper.isURL(answer.value)) {
    return {
      filename: '',
      url: answer.value
    }
  }

  return {
    filename: answer.value.filename || '',
    url: answer.value.url
  }
}

function rating(answer: Answer): string {
  return answer.value
}

function singleChoice(answer: Answer): string {
  const choice = answer.properties!.choices?.find(row => row.id === answer.value)
  return choice ? choice.label : ''
}

function multipleChoice(answer: Answer): string {
  return (answer.properties?.choices?.filter(row => answer.value.value.includes(row.id)) || [])
    .map(row => row.label)
    .concat([answer.value.other])
    .filter(row => helper.isValid(row))
    .join(', ')
}

function fullName(answer: Answer): FullNameValue {
  return answer.value
}

function address(answer: Answer): string {
  return [
    answer.value.address1,
    ',',
    answer.value.address2,
    answer.value.city,
    ',',
    answer.value.state,
    ',',
    answer.value.country,
    answer.value.zip
  ]
    .filter(Boolean)
    .join(' ')
}

function legalTerms(answer: Answer): string {
  return helper.isTrue(answer.value) ? 'Yes' : 'No'
}

function dateRange(answer: Answer): string {
  return [answer.value.start, answer.value.end].filter(Boolean).join(' - ')
}

function inputTable(answer: Answer): string {
  const columns = answer.properties?.tableColumns

  if (helper.isValidArray(columns)) {
    const result: string[] = []

    answer.value.forEach((values: Record<string, string>) => {
      if (helper.isValid(values)) {
        const row = columns!.map(column => values[column.id]).join(', ')
        result.push(row)
      }
    })

    return result.join('\n')
  }

  return ''
}

function payment(answer: Answer): string {
  const value = answer.value as ServerSidePaymentValue
  const price = Big(value.amount).div(100).toFixed(2)
  let result = CURRENCY_SYMBOLS[value.currency] + price

  if (helper.isValid(value.paymentIntentId)) {
    result = `Succeeded ${result}`
  } else {
    result = `Incomplete ${result}`
  }

  return result
}

export default {
  fileUpload,
  rating,
  singleChoice,
  multipleChoice,
  fullName,
  address,
  legalTerms,
  dateRange,
  inputTable,
  payment
}
