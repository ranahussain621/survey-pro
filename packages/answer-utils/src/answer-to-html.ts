import parser from './answer-parser'
import { Answer, FieldKindEnum, QUESTION_FIELD_KINDS } from '@heyform-inc/shared-types-enums'

export function answersToHtml(answers: Answer[]): string {
  const html = answers
    .map(answer => {
      const value = parseHtmlAnswer(answer)

      return `
<li>
  <h3>${answer.title}</h3>
  <p>${value}</p>
</li>
`
    })
    .join('')

  return `<ol>${html}</ol>`
}

function parseHtmlAnswer(answer: Answer): string {
  let value = ''

  if (QUESTION_FIELD_KINDS.includes(answer.kind)) {
    switch (answer.kind) {
      case FieldKindEnum.FILE_UPLOAD:
        const file = parser.fileUpload(answer)
        value = `${file.filename} (${file.url})`
        break

      case FieldKindEnum.RATING:
      case FieldKindEnum.OPINION_SCALE:
        value = parser.rating(answer)
        break

      case FieldKindEnum.YES_NO:
        value = parser.singleChoice(answer)
        break

      case FieldKindEnum.MULTIPLE_CHOICE:
      case FieldKindEnum.PICTURE_CHOICE:
        value = parser.multipleChoice(answer)
        break

      case FieldKindEnum.FULL_NAME:
        const name = parser.fullName(answer)
        value = `${name.firstName} ${name.lastName}`
        break

      case FieldKindEnum.ADDRESS:
        value = parser.address(answer)
        break

      case FieldKindEnum.LEGAL_TERMS:
        value = parser.legalTerms(answer)
        break

      case FieldKindEnum.DATE_RANGE:
        value = parser.dateRange(answer)
        break

      case FieldKindEnum.INPUT_TABLE:
        value = parser.inputTable(answer)
        break

      case FieldKindEnum.PAYMENT:
        value = parser.payment(answer)
        break

      default:
        value = answer.value?.toString()
    }
  }

  return value
}
