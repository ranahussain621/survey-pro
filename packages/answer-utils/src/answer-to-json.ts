import parser from './answer-parser'
import { Answer, FieldKindEnum } from '@heyform-inc/shared-types-enums'

interface AnswersToJsonOptions {
  plain?: boolean
}

export function answersToJson(
  answers: Answer[],
  options?: AnswersToJsonOptions
): Record<string, any> {
  const result: Record<string, any> = {}

  answers.forEach(answer => {
    result[answer.id] = parseJsonAnswer(answer, options?.plain)
  })

  return result
}

function parseJsonAnswer(answer: Answer, plain = false): any {
  let value: any

  switch (answer.kind) {
    case FieldKindEnum.FILE_UPLOAD:
      value = parser.fileUpload(answer)
      if (plain) {
        value = `${value.filename} (${value.url})`
      }
      break

    case FieldKindEnum.RATING:
    case FieldKindEnum.OPINION_SCALE:
      value = parser.rating(answer)
      break

    case FieldKindEnum.YES_NO:
      value = parser.singleChoice(answer)
      break

    case FieldKindEnum.MULTIPLE_CHOICE:
    case FieldKindEnum.PICTURE_CHOICE:
      value = parser.multipleChoice(answer)
      break

    case FieldKindEnum.FULL_NAME:
      value = parser.fullName(answer)
      if (plain) {
        value = `${value.firstName} ${value.lastName}`
      }
      break

    case FieldKindEnum.ADDRESS:
      value = answer.value
      if (plain) {
        value = parser.address(answer)
      }
      break

    case FieldKindEnum.LEGAL_TERMS:
      value = parser.legalTerms(answer)
      break

    case FieldKindEnum.DATE_RANGE:
      value = answer.value
      if (plain) {
        value = parser.dateRange(answer)
      }
      break

    case FieldKindEnum.INPUT_TABLE:
      value = answer.value
      if (plain) {
        value = parser.inputTable(answer)
      }
      break

    case FieldKindEnum.PAYMENT:
      value = answer.value
      if (plain) {
        value = parser.payment(answer)
      }
      break

    default:
      value = answer.value?.toString()
  }

  return value
}
