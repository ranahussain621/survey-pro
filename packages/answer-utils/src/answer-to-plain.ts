import parser from './answer-parser'
import { Answer, FieldKindEnum } from '@heyform-inc/shared-types-enums'
import { helper } from '@heyform-inc/utils'

export function answersToPlain(answers: Answer[]): string {
  return answers
    .map(answer => {
      const value = parsePlainAnswer(answer)
      return `${answer.title}\n${value}`
    })
    .join('\n\n')
}

export function parsePlainAnswer(answer: Answer, livePreview = false): string {
  let value: string

  switch (answer.kind) {
    case FieldKindEnum.FILE_UPLOAD:
      const file = parser.fileUpload(answer, livePreview)
      value = livePreview ? file.filename : `${file.filename} (${file.url})`
      break

    case FieldKindEnum.RATING:
    case FieldKindEnum.OPINION_SCALE:
      value = parser.rating(answer)
      break

    case FieldKindEnum.YES_NO:
      value = parser.singleChoice(answer)
      break

    case FieldKindEnum.MULTIPLE_CHOICE:
    case FieldKindEnum.PICTURE_CHOICE:
      value = parser.multipleChoice(answer)
      break

    case FieldKindEnum.FULL_NAME:
      const name = parser.fullName(answer)
      value = `${name.firstName} ${name.lastName}`
      break

    case FieldKindEnum.ADDRESS:
      value = parser.address(answer)
      break

    case FieldKindEnum.LEGAL_TERMS:
      value = parser.legalTerms(answer)
      break

    case FieldKindEnum.DATE_RANGE:
      value = parser.dateRange(answer)
      break

    case FieldKindEnum.INPUT_TABLE:
      value = parser.inputTable(answer)
      break

    case FieldKindEnum.PAYMENT:
      value = parser.payment(answer)
      break

    default:
      value = answer.value?.toString()
  }

  if (livePreview && helper.isEmpty(value)) {
    value = '_____'
  }

  return value
}
