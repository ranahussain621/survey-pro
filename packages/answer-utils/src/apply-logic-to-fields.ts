import { calculateAction } from './calculate-action'
import { validateRequiredField } from './validate'
import { validateCondition } from './validate-condition'
import {
  ActionEnum,
  AnswerValue,
  FieldKindEnum,
  FormField,
  Logic,
  NavigateAction,
  NumberCalculateAction,
  QUESTION_FIELD_KINDS,
  StringCalculateAction,
  Variable
} from '@heyform-inc/shared-types-enums'
import { helper } from '@heyform-inc/utils'

interface IFormField extends FormField {
  parent?: FormField
  isTouched?: boolean
}

export interface PickAndCalcFieldsResult {
  fields: IFormField[]
  variables: Record<string, AnswerValue>
}

function indexFields(fields: IFormField[]) {
  const parents = fields.filter(f => !f.parent && QUESTION_FIELD_KINDS.includes(f.kind))
  const children = fields.filter(f => f.parent && QUESTION_FIELD_KINDS.includes(f.kind))

  let index = 1
  const childrenIndexes: Record<string, number> = {}

  parents.forEach(f => {
    f.index = index++
  })

  children.forEach(f => {
    const parentId = f.parent!.id
    const parent = parents.find(p => p.id === parentId)

    if (!childrenIndexes[parentId]) {
      childrenIndexes[parentId] = 1
    }

    f.parent = parent
    f.index = childrenIndexes[parentId]++
  })
}

export function applyLogicToFields(
  fields?: IFormField[],
  logics?: Logic[],
  parameters?: Variable[],
  values?: Record<string, AnswerValue>
): PickAndCalcFieldsResult {
  const result: PickAndCalcFieldsResult = {
    fields: [],
    variables: {}
  }

  if (helper.isEmpty(fields)) {
    return result
  }

  if (helper.isValid(parameters)) {
    for (const variable of parameters!) {
      result.variables[variable.id] = variable.value
    }
  }

  if (helper.isEmpty(logics) || helper.isEmpty(values)) {
    // Add indexes to every field
    indexFields(fields!)

    return {
      ...result,
      fields: fields!.filter(f => f.kind !== FieldKindEnum.THANK_YOU)
    }
  }

  let index = 0

  while (index < fields!.length) {
    const field = fields![index]

    if (field.kind === FieldKindEnum.THANK_YOU) {
      break
    }

    let isNavigateValidated = false
    const logic = logics!.find(l => l.fieldId === field.id)

    // Add parent field if needed to present SubGroup fields
    if (field.parent && result.fields.findIndex(f => f.id === field.parent!.id) < 0) {
      result.fields.push(field.parent)
    }

    if (logic) {
      const { payloads } = logic
      const calculates = payloads.filter(p => p.action.kind === ActionEnum.CALCULATE)

      for (const calculate of calculates) {
        const { action, condition } = calculate
        const isValidated = validateCondition(field, condition, values)

        if (isValidated) {
          result.variables = calculateAction(
            action as NumberCalculateAction | StringCalculateAction,
            parameters,
            result.variables,
            values
          )
        }
      }

      const navigates = payloads.filter(p => p.action.kind === ActionEnum.NAVIGATE)

      // Find the first navigate which meets the jump rule
      for (const navigate of navigates) {
        const { action, condition } = navigate

        // Whether the jump logic in this field has been verified or not
        field.isTouched = validateCondition(field, condition, values)

        if (field.isTouched) {
          const jumpFieldId = (action as NavigateAction).fieldId
          const jumpIndex = fields!.findIndex(f => f.id === jumpFieldId)
          const isExists = !!result.fields.find(f => f.id === jumpFieldId)

          // Logic jump only works when sending the respondent forward in the form
          if (!isExists && jumpIndex > index) {
            index = jumpIndex
            isNavigateValidated = true

            result.fields.push(field)
            break
          }
        } else {
          field.isTouched = validateRequiredField(field, values)
        }
      }
    }

    if (!isNavigateValidated) {
      index += 1
      result.fields.push(field)
    }
  }

  // Add indexes to every field
  indexFields(result.fields)

  return result
}
