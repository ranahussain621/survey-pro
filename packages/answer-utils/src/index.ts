export * from './answer-to-html'
export * from './answer-to-api-object'
export * from './answer-to-plain'
export * from './answer-to-json'
export * from './field-values-to-answer'
export * from './fields-to-validate-rules'
export * from './validate'
export * from './html-utils'
export * from './flatten-fields'
export * from './validate-condition'
export * from './validate-payload'
export * from './apply-logic-to-fields'
export * from './consts'
export * from './helper'
