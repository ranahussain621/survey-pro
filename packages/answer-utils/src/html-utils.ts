import { helper } from '@heyform-inc/utils'
import { parse as html5Parse } from 'html5parser'
import { IAttribute } from 'html5parser/src/types'

interface HTMLWalkOptions {
  allowedTags?: string[]
  allowedBlockTags?: string[]
  allowedAttributes?: string[]
  plain?: boolean
  livePreview?: boolean
}

const ALLOWED_BLOCK_TAGS = ['div', 'h1', 'h2', 'h3', 'p', 'br']
const ALLOWED_TAGS = [
  'text',
  'span',
  'bold',
  'strong',
  'code',
  'a',
  'b',
  'i',
  'u',
  's',
  'mention',
  'variable',
  'hiddenfield'
]
const ALLOWED_ATTRIBUTES = [
  'href',
  'class',
  'data-mention',
  'data-variable',
  'data-hiddenfield',
  'contenteditable',
  'style'
]

function getAttributes(row: IAttribute[], allowedAttributes: string[] = []): Record<string, any> {
  const result: Record<string, any> = {}

  if (helper.isValidArray(row)) {
    row.forEach(a => {
      const name = a.name.value.toLowerCase()

      if (allowedAttributes.includes(name)) {
        result[name] = a.value?.value
      }
    })
  }

  return result
}

function walk(node: any, option: HTMLWalkOptions): any[] | string | undefined {
  let tag = (node.name || node.type).toLowerCase()
  const allowedTags = option.allowedTags!.concat(option.allowedBlockTags || ALLOWED_BLOCK_TAGS)

  if (allowedTags.includes(tag)) {
    const text = node.value

    if (tag === 'text') {
      return text
    }

    let attributes = getAttributes(node.attributes, option.allowedAttributes)

    if (tag === 'span') {
      const mentionId = attributes['data-mention']
      const variableId = attributes['data-variable']

      if (helper.isValid(mentionId)) {
        tag = 'mention'
        attributes = {
          id: mentionId
        }
      } else if (helper.isValid(variableId)) {
        tag = 'variable'
        attributes = {
          id: variableId
        }
      }
    }

    const schema: any[] = [tag]

    if (helper.isValidArray(node.body)) {
      const body = node.body!.map((child: any) => walk(child, option))

      if (helper.isValidArray(body)) {
        schema.push(body)
      }
    }

    if (helper.isValid(attributes)) {
      schema.push(attributes)
    }

    return schema
  }
}

// Purge illegal html tag and attributes
function purge(html: string, option?: HTMLWalkOptions): string {
  const schemas = parse(html, option)
  return serialize(schemas, option)
}

// HTML string to schema
function parse(html: string, option?: HTMLWalkOptions): any[] {
  if (helper.isEmpty(html)) {
    return []
  }

  const result = html5Parse(html)
  const customOption: HTMLWalkOptions = {
    allowedTags: ALLOWED_TAGS,
    allowedAttributes: ALLOWED_ATTRIBUTES,
    ...option
  }

  return result.map((node: any) => walk(node, customOption)).filter(Boolean)
}

// HTML schema to HTML string
function serialize(schemas?: any[], option?: HTMLWalkOptions): string {
  if (!helper.isValidArray(schemas)) {
    return ''
  }

  const customOption: HTMLWalkOptions = {
    allowedTags: ALLOWED_TAGS,
    allowedAttributes: ALLOWED_ATTRIBUTES,
    ...option
  }

  const allowedTags = customOption.allowedTags!.concat(
    customOption.allowedBlockTags || ALLOWED_BLOCK_TAGS
  )

  return schemas!
    .map(schema => {
      if (helper.isString(schema)) {
        return schema
      }

      if (!helper.isValidArray(schema)) {
        return ''
      }

      let [tag, body, attributes] = schema

      if (
        !allowedTags.includes(tag) ||
        // Only allows empty br node
        (helper.isEmpty(body) && helper.isEmpty(attributes) && tag !== 'br')
      ) {
        return ''
      }

      if (customOption.plain) {
        return serialize(body!, customOption)
      }

      if (tag === 'br') {
        return '<br />'
      }

      let property = ''

      if (helper.isValid(attributes)) {
        if (tag === 'mention') {
          attributes = {
            class: 'mention',
            contenteditable: 'false',
            'data-mention': attributes.id
          }
        } else if (tag === 'variable') {
          attributes = {
            class: 'variable',
            contenteditable: 'false',
            'data-variable': attributes.id
          }
        }

        property = Object.keys(attributes!)
          .filter(key => customOption.allowedAttributes!.includes(key))
          .map(key => ` ${key}="${attributes![key]}"`)
          .join('')
      }

      if (tag === 'mention' || tag === 'variable') {
        tag = 'span'

        // Placeholder for live preview
        if (customOption.livePreview) {
          body = ['_____']
        }
      }

      return `<${tag}${property}>${serialize(body!, customOption)}</${tag}>`
    })
    .join('')
}

// HTML to plain text
function plain(html: string, limit = 0) {
  if (helper.isEmpty(html)) {
    return ''
  }

  const result = html
    .replace(/<style[^<>]*>((?!<\/).)*<\/style>/gi, '')
    .replace(/<script[^<>]*>((?!<\/).)*<\/script>/gi, '')
    .replace(/<[^>]+>/g, '')
    .replace(/&nbsp;/g, ' ')
    .replace(/&quot;/g, '"')
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')

  if (limit > 0) {
    let sliced = result.slice(0, limit)

    if (result.length > limit) {
      sliced += '...'
    }

    return sliced
  }

  return result
}

export const htmlUtils = {
  parse,
  serialize,
  purge,
  plain
}
