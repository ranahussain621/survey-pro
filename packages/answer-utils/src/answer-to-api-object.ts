import { parsePlainAnswer } from './answer-to-plain'
import { Answer } from '@heyform-inc/shared-types-enums'

export function answersToApiObject(answers: Answer[]): Record<string, any> {
  const result: Record<string, any> = {}

  answers.forEach(answer => {
    const key = `(ID: ${answer.id}) ${answer.title}`
    result[key] = parsePlainAnswer(answer)
  })

  return result
}
