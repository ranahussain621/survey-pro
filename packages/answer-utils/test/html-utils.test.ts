import { htmlUtils } from '../src/html-utils'

const schema = [
  ['b', ['Make any website your Mac desktop wallpaper.&nbsp;']],
  ['div', [['b', [null]]]],
  [
    'div',
    [
      'Plash enables you to have a highly dynamic ',
      [
        'a',
        ['desktop wallpaper.'],
        {
          href: 'https://github.com',
          style: ''
        }
      ],
      ' You could display your favorite news site, Facebook feed, or a random beautiful scenery photo.'
    ]
  ]
]

test('serialize html', () => {
  expect(htmlUtils.serialize(schema, { allowedBlockTags: [] })).toMatchSnapshot()
})

test('serialize html with block tags', () => {
  expect(
    htmlUtils.serialize(schema, {
      allowedBlockTags: ['div', 'p']
    })
  ).toMatchSnapshot()
})

test('parse html', () => {
  expect(htmlUtils.parse(htmlUtils.serialize(schema, { allowedBlockTags: [] }))).toMatchSnapshot()
})

test('parse html with block tags', () => {
  const html = htmlUtils.serialize(schema, {
    allowedBlockTags: ['div', 'p']
  })

  expect(
    htmlUtils.parse(html, {
      allowedBlockTags: ['div', 'p']
    })
  ).toMatchSnapshot()
})

test('plain html', () => {
  expect(
    htmlUtils.plain(
      '<p><strong>Programmer\'s guide</strong> about how to <a href="https://github.com">cook at home</a></p>'
    )
  ).toBe("Programmer's guide about how to cook at home")
})

test('plain html with limit', () => {
  expect(
    htmlUtils.plain(
      '<p><strong>Programmer\'s guide</strong> about how to <a href="https://github.com">cook at home</a></p>',
      20
    )
  ).toBe("Programmer's guide a...")
})

test('plain html with large limit', () => {
  expect(
    htmlUtils.plain(
      '<p><strong>Programmer\'s guide</strong> about how to <a href="https://github.com">cook at home</a></p>',
      100
    )
  ).toBe("Programmer's guide about how to cook at home")
})
