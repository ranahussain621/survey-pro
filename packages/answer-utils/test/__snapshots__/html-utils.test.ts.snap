// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse html 1`] = `
Array [
  Array [
    "b",
    Array [
      "Make any website your Mac desktop wallpaper.&nbsp;",
    ],
  ],
]
`;

exports[`parse html with block tags 1`] = `
Array [
  Array [
    "b",
    Array [
      "Make any website your Mac desktop wallpaper.&nbsp;",
    ],
  ],
  Array [
    "div",
    Array [
      Array [
        "b",
      ],
    ],
  ],
  Array [
    "div",
    Array [
      "Plash enables you to have a highly dynamic ",
      Array [
        "a",
        Array [
          "desktop wallpaper.",
        ],
        Object {
          "href": "https://github.com",
          "style": "",
        },
      ],
      " You could display your favorite news site, Facebook feed, or a random beautiful scenery photo.",
    ],
  ],
]
`;

exports[`serialize html 1`] = `"<b>Make any website your Mac desktop wallpaper.&nbsp;</b>"`;

exports[`serialize html with block tags 1`] = `"<b>Make any website your Mac desktop wallpaper.&nbsp;</b><div><b></b></div><div>Plash enables you to have a highly dynamic <a href=\\"https://github.com\\" style=\\"\\">desktop wallpaper.</a> You could display your favorite news site, Facebook feed, or a random beautiful scenery photo.</div>"`;
